packages/asset/contracts/AssetCreate.sol

```diff
@@ -0,0 +1,336 @@
+        );
+        // all arrays must be same length
+        require(
+            tiers.length == amounts.length &&
+                amounts.length == metadataHashes.length,
+            "Arrays must be same length"
+        );
+
+        uint256[] memory tokenIds = new uint256[](tiers.length);
+        uint256[] memory tiersToBurn = new uint256[](tiers.length);
+        for (uint256 i = 0; i < tiers.length; i++) {
+            uint16 creatorNonce = getCreatorNonce(creator);
+            tiersToBurn[i] = tiers[i];
+            tokenIds[i] = TokenIdUtils.generateTokenId(
+                creator,
+                tiers[i],
+                creatorNonce,
+                revealed[i] ? 1 : 0
+            );
+        }
+
+        catalystContract.burnBatchFrom(creator, tiersToBurn, amounts);
+
+        assetContract.mintBatch(creator, tokenIds, amounts, metadataHashes);
+        emit AssetBatchMinted(
+            creator,
+            tokenIds,
+            tiers,
+            amounts,
+            metadataHashes
+        );
+    }
+
+    /// @notice Create special assets, like TSB exclusive tokens
+    /// @dev Only callable by the special minter
+    /// @param signature A signature generated by TSB
+    /// @param tier The tier of the asset to mint
+    /// @param amount The amount of the asset to mint
+    /// @param metadataHash The metadata hash of the asset to mint
+    function createSpecialAsset(
+        bytes memory signature,
+        uint8 tier,
+        uint256 amount,
+        bool revealed,
+        string calldata metadataHash
+    ) external onlyRole(SPECIAL_MINTER_ROLE) {
+        address creator = _msgSender();
+        require(
+            authValidator.verify(
+                signature,
+                _hashMint(creator, tier, amount, revealed, metadataHash)
+            ),
+            "Invalid signature"
+        );
+
+        uint16 creatorNonce = getCreatorNonce(creator);
+        uint256 tokenId = TokenIdUtils.generateTokenId(
+            creator,
+            tier,
+            creatorNonce,
+            revealed ? 1 : 0
+        );
+
+        assetContract.mint(creator, tokenId, amount, metadataHash);
+        emit SpecialAssetMinted(creator, tokenId, amount, metadataHash);
+    }
+
+    /// @notice Create a bridged asset
+    /// @dev Only callable by the bridge minter
+    /// @dev Bridge contract should keep track of the tokenIds generated so that copies of the same asset on L1 are minted with the same tokenIds on L2
+    /// @param recipient The address of the creator
+    /// @param tokenId Id of the token to mint
+    /// @param amount The amount of copies to mint
+    /// @param metadataHash The metadata hash of the asset
+    function createBridgedAsset(
+        address recipient,
+        uint256 tokenId,
+        uint256 amount,
+        string calldata metadataHash
+    ) external onlyRole(BRIDGE_MINTER_ROLE) {
+        assetContract.mint(recipient, tokenId, amount, metadataHash);
```

Comment:
How to avoid clashes with token IDs generated by this contract?

Reply:
That's right, we might encounter a potential clas when a creator mints a new asset at the same block as an asset of his from L1 is being bridged.

We need to make sure the token id generated by the bridge is different from the others.

Hmm I was thinking we might add a flag to the tokenId to indicate whether the asset has been bridged?
What do you think?

Reply:
Will this change to tokenId be made in a separate PR? @wojciech-turek @rishabh0x00 ?

Reply:
Change to the tokenId to include the bridged bit was added in this PR so we should be good.

Reply:
Added newest commit with a function for the bridge contract to get and increment the creatorNonce.

The bridge contract should be responsible to fetch it only if bridging a token for the first time after that they would be increasing the supply of an already existing token.